File: packages/python/qraudio/codec/afskModem.py
from __future__ import annotations

import math

from .envelope import applyFade


def tonesToSamples(
    tones: list[int],
    *,
    sample_rate: float,
    baud: float,
    mark_freq: float,
    space_freq: float,
    level_db: float,
    fade_ms: float,
) -> list[float]:
    samples_per_bit = sample_rate / baud
    total_samples = math.ceil(len(tones) * samples_per_bit)
    out: list[float] = [0.0] * total_samples
    amplitude = 10 ** (level_db / 20.0)

    phase = 0.0
    sample_index = 0
    boundary = samples_per_bit

    for tone in tones:
        freq = mark_freq if tone == 1 else space_freq
        phase_step = (2 * math.pi * freq) / sample_rate
        while sample_index < boundary and sample_index < total_samples:
            phase += phase_step
            if phase > math.pi * 2:
                phase -= math.pi * 2
            out[sample_index] = math.sin(phase) * amplitude
            sample_index += 1
        boundary += samples_per_bit

    applyFade(out, sample_rate, fade_ms)
    return out


def demodAfsk(
    samples: list[float],
    sample_rate: float,
    baud: float,
    offset: int,
    mark_freq: float,
    space_freq: float,
) -> list[int]:
    samples_per_bit = sample_rate / baud
    tones: list[int] = []

    start = offset
    boundary = start + samples_per_bit

    while boundary <= len(samples):
        end = math.floor(boundary)
        length = end - start
        if length <= 1:
            start = end
            boundary += samples_per_bit
            continue
        mark_energy = goertzel(samples, start, length, mark_freq, sample_rate)
        space_energy = goertzel(samples, start, length, space_freq, sample_rate)
        tones.append(1 if mark_energy >= space_energy else 0)
        start = end
        boundary += samples_per_bit

    return tones


def goertzel(
    samples: list[float],
    start: int,
    length: int,
    freq: float,
    sample_rate: float,
) -> float:
    omega = (2 * math.pi * freq) / sample_rate
    coeff = 2 * math.cos(omega)
    s0 = 0.0
    s1 = 0.0
    s2 = 0.0
    end = start + length
    for i in range(start, end):
        s0 = samples[i] + coeff * s1 - s2
        s2 = s1
        s1 = s0
    return s1 * s1 + s2 * s2 - coeff * s1 * s2

File: packages/js/src/core/codec/afskModem.ts
import { applyFade } from "./envelope.js";

interface ToneOptions {
  sampleRate: number;
  baud: number;
  markFreq: number;
  spaceFreq: number;
  levelDb: number;
  fadeMs: number;
}

export function tonesToSamples(tones: number[], options: ToneOptions): Float32Array {
  const { sampleRate, baud, markFreq, spaceFreq, levelDb, fadeMs } = options;
  const samplesPerBit = sampleRate / baud;
  const totalSamples = Math.ceil(tones.length * samplesPerBit);
  const out = new Float32Array(totalSamples);
  const amplitude = Math.pow(10, levelDb / 20);

  let phase = 0;
  let sampleIndex = 0;
  let boundary = samplesPerBit;

  for (let bitIndex = 0; bitIndex < tones.length; bitIndex += 1) {
    const freq = tones[bitIndex] === 1 ? markFreq : spaceFreq;
    const phaseStep = (2 * Math.PI * freq) / sampleRate;
    while (sampleIndex < boundary && sampleIndex < totalSamples) {
      phase += phaseStep;
      if (phase > Math.PI * 2) {
        phase -= Math.PI * 2;
      }
      out[sampleIndex] = Math.sin(phase) * amplitude;
      sampleIndex += 1;
    }
    boundary += samplesPerBit;
  }

  applyFade(out, sampleRate, fadeMs);
  return out;
}

export function demodAfsk(
  samples: Float32Array,
  sampleRate: number,
  baud: number,
  offset: number,
  markFreq: number,
  spaceFreq: number
): number[] {
  const samplesPerBit = sampleRate / baud;
  const tones: number[] = [];

  let start = offset;
  let boundary = start + samplesPerBit;

  while (boundary <= samples.length) {
    const end = Math.floor(boundary);
    const len = end - start;
    if (len <= 1) {
      start = end;
      boundary += samplesPerBit;
      continue;
    }
    const markEnergy = goertzel(samples, start, len, markFreq, sampleRate);
    const spaceEnergy = goertzel(samples, start, len, spaceFreq, sampleRate);
    tones.push(markEnergy >= spaceEnergy ? 1 : 0);
    start = end;
    boundary += samplesPerBit;
  }

  return tones;
}

function goertzel(
  samples: Float32Array,
  start: number,
  length: number,
  freq: number,
  sampleRate: number
): number {
  const omega = (2 * Math.PI * freq) / sampleRate;
  const coeff = 2 * Math.cos(omega);
  let s0 = 0;
  let s1 = 0;
  let s2 = 0;
  const end = start + length;
  for (let i = start; i < end; i += 1) {
    s0 = samples[i] + coeff * s1 - s2;
    s2 = s1;
    s1 = s0;
  }
  return s1 * s1 + s2 * s2 - coeff * s1 * s2;
}
